### PR TITLE
test: close alive mutations in 7 orchestration service specs

### LINKED ISSUE #117: test: close 304 alive mutations in orchestration services (91.14% → 100%)
## Background

Focused mutant run across 8 high-value orchestration service classes (71 subjects, 3,435 mutations):

```
Mutations: 3,435  |  Kills: 3,131  |  Alive: 304  |  Timeouts: 25
Coverage:  91.14%
```

Notable: `Orchestration::PipelineRunner` (54 tests, 11 subjects) came back **fully killed — 0 alive mutations**. No work needed there.

The 304 remaining alive mutations cluster into 7 classes below, ordered by severity.

---

## 1. `NormalizeActionRunFailure` — highest priority

**Only 3 tests cover 29 methods.** Every conditional branch, error-classification path, sanitization rule, and string-manipulation edge case is invisible to the suite. This is the most important class to address.

Representative alive mutations:

```diff
# #call — mutant can hardcode the branch condition
-  if invalid_model_output?
+  if true
     return invalid_model_output_result
   end
```

```diff
# #extract_parsed_error — hash access vs fetch not distinguished
-    parsed["error"] || parsed
+    parsed.fetch("error") || parsed
```

**Fix**: Expand `spec/services/orchestration/normalize_action_run_failure_spec.rb` to cover each error-classification path in isolation: `provider_http_error?`, `transport_error?`, `invalid_model_output?` each returning true/false independently; `sanitize_string`/`sanitize_value` with sensitive keys; `truncate` at boundaries; `stringify` with non-string inputs.

---

## 2. `SchemaValidator#validate_node!` — ~32 alive

20 tests exist but `validate_node!` has 25+ alive mutations. The recursive type-dispatch logic is barely exercised.

```diff
# schema["type"] vs schema.fetch("type") not caught
-  case schema["type"]
+  case schema.fetch("type")
```

**Fix**: Add tests that cover each `schema["type"]` branch (`"object"`, `"array"`, `"string"`, `"number"`, `"boolean"`, `nil`/absent) with both valid and invalid values, missing required keys, and nested schemas.

---

## 3. `IngestionExecutor` — ~39 alive across 6 class methods

17 tests cover ingestion, but transformation operations have gaps:

```diff
# apply_merge_by_index — .to_s coercion on source key not verified
-    source = op.fetch("source").to_s
+    source = op.fetch("source")
```

```diff
# apply_rename — 15 more alive for similar reasons
```

**Fix** (`spec/services/orchestration/ingestion_executor_spec.rb`): Add tests that assert `.to_s` coercion on op keys actually matters (e.g., pass an integer key), verify `apply_filter_by_ids` with no matching IDs, `apply_merge_by_index` when arrays are different lengths, `dig_path` with non-existent segments, `item_id` with the `id_field` option.

---

## 4. `RuntimeAgentBuilder` — ~46 alive + neutral failures

Two subjects (`#runtime_chat`, `#snapshot`) reported **neutral failures** — the original test failed due to `SQLite3::BusyException: database is locked` during parallel killfork execution. These are not coverage gaps but a test isolation issue: the spec uses `create(:orchestration_agent)` in a `let` that races when multiple killforks share the same SQLite database.

Remaining alive mutations (excluding neutrals):

```diff
# #resolved_tools — @tool_classes guard not tested
-  if @tool_classes.present?
-    return @tool_classes
-  end

# configured_tools&.map — nil guard not tested
-  configured_tools&.map do |tool|
+  configured_tools.map do |tool|

# tool.is_a?(Class) branch — Class subclass not tested
-  if tool.is_a?(Class)
+  if false
```

**Fix (two parts)**:
1. **SQLite locking**: Wrap `runtime_agent_builder_spec.rb` database setup in `DatabaseCleaner` or use `build_stubbed`/`instance_double` to avoid hitting the database in killforks.
2. **Coverage gaps**: Add tests for `resolved_tools` with `@tool_classes` pre-set (the override path), with `configured_tools: nil` (the safe-navigation path), and with a tool that is already a Class constant.

---

## 5. `QueryExecutor` — ~5 alive

```diff
# dig_path — same nil-guard issue as InputMappingResolver
-  if node.nil?
-    break nil
-  end
```

**Fix**: Add a test where an intermediate path node is `nil`.

---

## 6. `InputMappingUpdater` — ~3 alive

Minor gaps around the `call` delegation path.

**Fix**: Add a test that verifies `InputMappingUpdater.call(...)` routes through the instance method correctly, and one where `input_mapping` is empty.

---

## 7. `OutputKeyDeriver` — ~2 alive

```diff
# #call — one more branch variant not covered
```

**Fix**: Verify behaviour when the action has no output key set (nil case).

---

## Acceptance criteria

- [ ] `NormalizeActionRunFailure` coverage reaches 100%
- [ ] `SchemaValidator#validate_node!` coverage reaches 100%
- [ ] `IngestionExecutor` transformation methods reach 100%
- [ ] `RuntimeAgentBuilder` SQLite locking resolved and coverage reaches 100%
- [ ] `QueryExecutor`, `InputMappingUpdater`, `OutputKeyDeriver` gaps closed
- [ ] `bundle exec mutant run` (focused config below) exits 0

## How to reproduce

```yaml
# .mutant.yml (temporary override — restore after)
integration: rspec
usage: opensource
jobs: 4
includes: [app/services, spec]
requires: [rspec/core, mutant_helper]
matcher:
  subjects:
    - Orchestration::IngestionExecutor
    - Orchestration::NormalizeActionRunFailure
    - Orchestration::PipelineRunner
    - Orchestration::SchemaValidator
    - Orchestration::OutputKeyDeriver
    - Orchestration::InputMappingUpdater
    - Orchestration::QueryExecutor
    - Orchestration::RuntimeAgentBuilder
```

```bash
bundle exec mutant run
```

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### PR DESCRIPTION
**Pull request by alaexeyshustov**
## Summary

- Expanded 7 orchestration service specs from 71 → 111 examples to kill the 304 alive mutations identified in issue #117
- Fixed RuntimeAgentBuilder SQLite locking by providing chat in test setup
- Added targeted tests for each conditional guard, sanitization path, coercion, and edge case called out in the issue

Closes #117

## Test plan

- [x] All 7 new spec files pass (111 examples, 0 failures)
- [x] Full suite passes (1884 examples, 0 failures)
- [x] RuboCop passes with no offenses

🤖 Generated with [Claude Code](https://claude.com/claude-code)
   Tokens used: input=65, output=74254, cache_read=5234548, cache_write=124872

### CODE REVIEW

**Comment by alaexeyshustov** [comment]
## Code Review

### 🔒 Security
No issues. Test data with fake credentials (`"sk-secret"`, `"Bearer sk-abc123def456ghi"`) is intentional fixture data used to verify redaction logic — it passes the sensitive-key filter and is confirmed redacted by the assertions.

### 📐 Code Quality

**medium — `input_mapping_updater_spec.rb` (step_result nil context)**
The context *"when the step_action does not appear in the validator results"* creates `other_step` inside the same `pipeline` as the outer `let(:pipeline)`. Because `Pipeline::Validator.call(pipeline)` validates all step_actions in that pipeline, `step_result` may not actually be `nil` — making the test equivalent to the "empty mapping" test above it. If the intent was to exercise the `step_result&.errors` safe-navigation guard, the `other_step_action` should be created in a separate pipeline. Minor mismatch between test name and setup.

**low — `normalize_action_run_failure_spec.rb` (repeated `let(:model)`)**
Several new top-level contexts each redeclare `let(:model) { "gpt-4.1-mini" }`. Could be moved to a shared `context "with a provider HTTP error"` or use a file-scoped `let`. Not a bug, just mild duplication.

**low — `runtime_agent_builder_spec.rb` (four identical empty-policy `let` blocks)**
The four contexts that test blank/nil policy fields each define an identical `let(:policy)` override. A single shared context would remove the repetition while keeping the test intent clear.

### ⚙️ Functionality

All assertion chains are logically sound:
- `Data.define(:status, :body)` correctly satisfies the `respond_to?(:response)` + `.present?` + `.status`/`.body` checks in `provider_http_error?` and its helpers.
- `allow(chat).to receive(:with_model).and_return(chat)` blocks the real `Models.resolve` API call while still recording the invocation for `have_received`.
- Symbol-key coercion tests (`:emails`, `:id`, etc.) correctly distinguish `hash.fetch(:emails, [])` (miss) from `hash.fetch("emails", [])` (hit) — each test exercises the `.to_s` call.
- `change(Chat, :count).by(1)` in the no-chat test is accurate: only `Chat.create!` adds a Chat row; `with_model`/`with_instructions` update or create related records (Model, Message) but not Chats.

### Summary

The PR is mergeable. Top 3 items to consider addressing:
1. **Verify the `InputMappingUpdater` nil-step_result context** actually achieves `step_result = nil` (create `other_step` in a fresh pipeline instead of re-using `pipeline`).
2. **Deduplicate the four identical empty-policy `let` blocks** in `runtime_agent_builder_spec.rb` into a shared context.
3. **Move the repeated `let(:model) { "gpt-4.1-mini" }` declarations** to a shared outer context in `normalize_action_run_failure_spec.rb`.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

File: spec/services/orchestration/runtime_agent_builder_spec.rb [hunk 1]
```diff
@@ -18,6 +20,98 @@
...
 
+    it 'uses the provided chat' do
+      expect(builder.build.chat).to eq(chat)
+    end
+
+    it 'applies the policy model to the agent' do
+      allow(chat).to receive(:with_model).and_return(chat)
+      builder.build
+      expect(chat).to have_received(:with_model).with("mistral-large-latest")
+    end
```
[the remark this answers is not in the document]
Comment (alaexeyshustov) [reply spec/services/orchestration/runtime_agent_builder_spec.rb:33]: Fixed in 64ee2a8. Added an inline comment in the spec explaining the delegation chain:  uses , so  calls . We stub the chat directly rather than the agent to avoid creating test doubles.

File: spec/services/orchestration/input_mapping_updater_spec.rb [hunk 2]
```diff
@@ -104,5 +104,31 @@
...
+    end
+
+    context "when the step_action does not appear in the validator results" do
+      it "treats missing step_result as no errors and no warnings" do
+        other_step = create(:orchestration_step, pipeline: pipeline)
+        other_action = create(:orchestration_action, name: "Other Action")
+        other_step_action = create(:orchestration_step_action, step: other_step, action: other_action, output_key: "other")
+
+        result = described_class.call(
+          step_action: other_step_action,
```
[the remark this answers is not in the document]
Comment (alaexeyshustov) [reply spec/services/orchestration/input_mapping_updater_spec.rb]: Fixed in 64ee2a8. Replaced the "other step in the same pipeline" setup with `allow(Orchestration::Pipeline::Validator).to receive(:call).and_return([])`. Since Validator always emits a StepResult for every step_action in the pipeline, stubbing to an empty array is the only reliable way to exercise the `step_result = nil` code path.

File: spec/services/orchestration/runtime_agent_builder_spec.rb [hunk 3]
```diff
@@ -1,7 +1,9 @@
 require 'rails_helper'
 
 RSpec.describe Orchestration::RuntimeAgentBuilder do
-  subject(:builder) { described_class.new(policy: policy) }
+  subject(:builder) { described_class.new(policy: policy, chat: chat) }
+
+  let!(:chat) { create(:chat) }
```
[the remark this answers is not in the document]
Comment (alaexeyshustov) [reply spec/services/orchestration/runtime_agent_builder_spec.rb]: Fixed in 64ee2a8. Changed to lazy `let(:chat)` so the DB write only happens for examples that actually reference `chat`, preventing an extra write per example in the "when no chat is provided" context.
